pkg/diagnostics/grpc_tracing.go

```diff
@@ -51,13 +51,31 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
 			spanKind = trace.WithSpanKind(trace.SpanKindClient)
 		}
 
-		ctx, span = trace.StartSpanWithRemoteParent(ctx, spanName, sc, sampler, spanKind)
+		ctx, span = trace.StartSpanWithRemoteParent(ctx, info.FullMethod, sc, sampler, spanKind)
+		if span.SpanContext().IsSampled() {
+			// users can add dapr- prefix if they want to see the header values in span attributes.
+			prefixedMetadata = userDefinedMetadata(ctx)
+			reqSpanAttr = spanAttributesMapFromGRPC(appID, req, info.FullMethod)
+		}
+
 		resp, err := handler(ctx, req)
 
-		addSpanMetadataAndUpdateStatus(ctx, span, info.FullMethod, appID, req, false)
+		if span.SpanContext().IsSampled() {
+			// Populates dapr- prefixed header first
+			for key, value := range reqSpanAttr {
+				prefixedMetadata[key] = value
+			}
+			AddAttributesToSpan(span, prefixedMetadata)
+
+			// Correct the span name based on API.
+			if sname, ok := reqSpanAttr[daprAPISpanNameInternal]; ok {
+				span.SetName(sname)
+			}
+		}
 
 		// Add grpc-trace-bin header for all non-invocation api's
-		if info.FullMethod != "/dapr.proto.runtime.v1.Dapr/InvokeService" {
+		target := fmt.Sprintf("/%s/InvokeService", runtimev1pb.Dapr_ServiceDesc.ServiceName)
```

Comment:
I would keep the constant in favor of avoiding string allocation.

Reply:
Sorry for force push :(